### PR TITLE
Redirect io

### PR DESCRIPTION
**Pull request by akleb**
## Purpose

This PR moves the multipoint `redirectIO` function to baseclasses. It also includes a new function that allows the user to use a `with` block that only redirects the IO during the `with` block. Both of these functions also have increased capability to split the `stderr` and `stdout` into different files if desired.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing

I added tests to the new `test_redirectIO.py` file. I think these are failing because `testflo` is capturing output and messing with the file streams before the tests are run. I am not sure how we can get around. I have tested this with my own scripts locally and it seems to work. 

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by akleb** [comment]
I have been using this with both pyHyp and ADflow and they both are redirecting IO properly in a `with` block, something like this:

```
f_stream = open(os.path.join(os.path.dirname(output), self.getOption("meshLog")), "w")
with redirectIO.redirectingIO(f_stream):
    hyp = pyHyp(options=options)
    hyp.run()
    hyp.writeCGNS(output + "_L0.cgns")
print("Saying something back in terminal instead of f_stream")
```
The `redirectingIO` function closes the `f_stream` after it is finished with it

**Review by marcomangano — COMMENTED** [review]
We should bump the version here as well

File: baseclasses/utils/redirectIO.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    Parameters
+    ----------
+    f_out : file
+        A file stream to redirect stdout to
+
+    f_err : file
+        A file stream to redirect stderr to, if none is specified it is set to `f_out`
+    """
+
+    if not f_err:
```
Comment (sseraj) [inline baseclasses/utils/redirectIO.py:26]: I would change this to `if f_err is None` to be more explicit

File: baseclasses/utils/redirectIO.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    Parameters
+    ----------
+    f_out : file
+        A file stream that stdout should be redirected to
+
+    f_err : file
+        A file stream that stderr should be redirected to
+    """
+
+    if not f_err:
```
Comment (sseraj) [inline baseclasses/utils/redirectIO.py:71]: `if f_err is None` here as well

File: baseclasses/utils/redirectIO.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    """
+    This function redirects stdout/stderr to the given file handle.
+
+    Parameters
+    ----------
+    f_out : file
+        A file stream to redirect stdout to
+
+    f_err : file
+        A file stream to redirect stderr to, if none is specified it is set to `f_out`
```
Comment (sseraj) [inline baseclasses/utils/redirectIO.py:23]: This should be two sentences

File: baseclasses/utils/redirectIO.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    ...     CFDSolver(AeroProblem(**apOptions)
+    >>> print("Printing some more information to terminal")
+
+    Parameters
+    ----------
+    f_out : file
+        A file stream that stdout should be redirected to
+
+    f_err : file
+        A file stream that stderr should be redirected to
```
Comment (sseraj) [inline baseclasses/utils/redirectIO.py:68]: This should be consistent with the `redirectIO` docstring

**Review by sseraj — COMMENTED** [review]
Looks pretty good. Some minor comments

File: baseclasses/utils/redirectIO.py (hunk 2 above)
Comment (akleb) [reply baseclasses/utils/redirectIO.py:71]: so many word

**Review by sseraj — APPROVED** [review]

**Review by marcomangano — APPROVED** [review]

**Pull request merged**